### ISSUE TITLE
Command-line arguments to Windows exe not quoted when they contain tabs

### ISSUE THREAD

**Issue by cspotcode** [Microsoft/WSL#2164]
### Your Windows build number: 
`Microsoft Windows [Version 10.0.15063]`

### What you're doing and what's happening:

When I invoke a Windows executable from WSL and one of the command-line arguments contains a tab, WSL is not wrapping it in double-quotes.  Windows executables usually split command-line args on tabs.

```
# from bash shell
# Argument with spaces is quoted
$ /mnt/c/EchoArgs.exe "$(printf 'a  b')"
Arg 0 is <a  b>

Command line:
"C:\EchoArgs.exe" "a  b"

# Argument with tabs is not quoted
$ /mnt/c/EchoArgs.exe "$(printf 'a\t\tb')"
Arg 0 is <a>
Arg 1 is <b>

Command line:
"C:\EchoArgs.exe" a         b
```

EchoArgs.exe is from Pscx; real path is `C:\Program Files (x86)\PowerShell Community Extensions\Pscx3\Pscx\Apps\EchoArgs.exe` but I truncated it above for readability.

### What's wrong / what should be happening instead:
WSL should be wrapping args that contain tabs in double-quotes like it does for args containing spaces.

It appears that double-quotes are not necessary for `\v`, `\r`, `\n`, or `\f` characters, so those don't cause any problems.

Sorry if this is a duplicate; I couldn't find a preexisting issue.

**Comment by sunjoong** [comment]
@cspotcode - Hmm... what if use external arguments that make block? I could not sure... but...
> EchoArgs.exe  '"' "$(printf 'a\t\tb')" '"'

Or
> EchoArgs.exe  '\"' "$(printf 'a\t\tb')" '\"'

This reminded me #2069 to make this elisp function;
```
(defun browse-url-edge (url &optional new-window)
  (shell-command
   (concat "/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe "
           "start microsoft-edge:'\\\"" url "\\\"'")))
```
But... I don't have EchoArgs.exe and EchoArgs.exe looked like to need install dot.NET 3, so I could not test deeper.

**Comment by cspotcode** [comment]
You can create your own `EchoArgs.exe`.  Just compile a simple C program that calls `printf()` for each entry in the `argv` array.  You can also write it in Python, Node, or Java.

Some of the solutions you're attempting are being affected by bash's argument splitting, so you should modify your tests to output the argv array in Linux first, and then pass it to EchoArgs.exe, so you can see the difference.  Otherwise you'll be accidentally splitting arguments before passing them to the Windows exe.

It's possible to manually wrap arguments in double-quotes, but this must be done *only* for arguments that contain tabs and not spaces.  It *cannot* be done for arguments containing spaces.  Essentially, the workaround must understand WSL's inconsistent quoting behavior.  I don't think this is what the WSL developers intended, which is why I reported this behavior as a bug.

**Comment by sunjoong** [comment]
@cspotcode - Ah... I think you mean...
```
sunjoong@SUNJOONG-DESKTOP ~/temp $ /mnt/c/msys64/home/sunjoong/echoargs_mingw.exe "$(printf 'a\t\tb')"
This program was called with 3 arguments.
The name of this program is [C:\msys64\home\sunjoong\echoargs_mingw.exe]
argv[0] = C:\msys64\home\sunjoong\echoargs_mingw.exe
argv[1] = a
argv[2] = b
sunjoong@SUNJOONG-DESKTOP ~/temp $
```
should same as
```
sunjoong@SUNJOONG-DESKTOP ~/temp $ ./echoargs_linux "$(printf 'a\t\tb')"
This program was called with 2 arguments.
The name of this program is [./echoargs_linux]
argv[0] = ./echoargs_linux
argv[1] = a             b
sunjoong@SUNJOONG-DESKTOP ~/temp $
```

**Comment by TSlivede** [comment]
The problem is not only about tabs, it's the whole "convert argv array to single CommandLine string for windows". AFAIK almost all Windows executables follow [these rules](https://msdn.microsoft.com/en-us/library/17w5ykft.aspx), so WSL should build the CommandLine accordingly:   
- If an argument contains whitespace (including tab) it should be surrounded by quotes.
- If an argument contains `"` those should be escaped with backslashes (see https://github.com/Microsoft/BashOnWindows/issues/1625)
- etc.

Even many applications that don't follow these rules could benefit from this: AFAIK cygwin's slightly incompatible commandline parsing (or even [`bash.exe`'s current strange commandline parsing](https://github.com/Microsoft/BashOnWindows/issues/1746) ) would profit from this change.
(Many of these apps would profit from quoting arguments not only if they contain space or tab but also if they contain `'` - this is perfectly compatible for apps that do follow [MS rules](https://msdn.microsoft.com/en-us/library/17w5ykft.aspx) )

A similar issue exists for powershell (https://github.com/PowerShell/PowerShell/issues/1995). [Fixing this in powershell](https://github.com/PowerShell/PowerShell-RFC/pull/90) requires many thoughts about backward compatibility.  
For WSL fixing should be easier, because WSL is still beta.

**Comment by benhillis** [comment]
Thanks again for reporting this issue, I have drafted a fix that resolves this and #1625.

**Comment by sunilmut** [comment]
This should be fixed in Fall Creators Update or any build > [16273](https://msdn.microsoft.com/en-us/commandline/wsl/release-notes#build-16273)
